Add manifest and files functions

matthewrsj opened the pull request:
This PR adds functions for file flags and manifest reading as well as tests for each.

Tests can be run with `make swupd.test` from the root of the repo.

matthewrsj committed:
swupd: add file flag library functions

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj committed:
swupd/files: add tests for file flag functions

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj committed:
swupd: add manifest read functions

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj committed:
swupd/manifest: add tests for manifest read functions

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

icarus-sparry commented on src/swupd/files.go:
```diff
@@ -0,0 +1,130 @@
+	Version     uint32
+
+	// flags
+	Type        string
```
Most (all) of Type, Status, Modifier should not be strings by their own type.

icarus-sparry commented on src/swupd/files.go:
```diff
@@ -0,0 +1,130 @@
+	Rename      bool
+
+	// renames
+	RenameScore uint16
```
Probably premature optimization to make RenameScore uint16 ratherthanint.

icarus-sparry commented on src/swupd/files.go:
```diff
@@ -0,0 +1,130 @@
+// .: none (may be deleted/ghosted)
+func setFileTypeFromFlag(flag string, f *File) error {
+	switch flag {
+	case "F":
```
These strings should be the associated "enum" type values,rather than strings.

icarus-sparry commented on src/swupd/files.go:
```diff
@@ -0,0 +1,130 @@
+// L: link
+// M: manifest
+// .: none (may be deleted/ghosted)
+func setFileTypeFromFlag(flag string, f *File) error {
```
If the flag is a character we should say so, rather than pretending it is a string.

icarus-sparry commented on src/swupd/files.go:
```diff
@@ -0,0 +1,130 @@
+
+	var err error
+	// set file type
+	if err = setFileTypeFromFlag(string(flags[0]), f); err != nil {
```
The fact that you have to cast this to a string is a strong hint that this is wrong.

icarus-sparry commented on src/swupd/files_test.go:
```diff
@@ -0,0 +1,178 @@
+	}
+
+	// space is never valid
+	t.Run("\" \"", func(t *testing.T) {
```
Go has more than one type of quote, so you can usually use a different one rather than needing to escae character.

icarus-sparry commented on src/swupd/files_test.go:
```diff
@@ -0,0 +1,178 @@
+		".": "",
+	}
+
+	for flag, ftype := range flags {
```
This is not a typical style for a test. See [wiki](https://github.com/golang/go/wiki/TableDrivenTests) or [cheney's blog](https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go) for a more normal style.

icarus-sparry commented on src/swupd/manifest.go:
```diff
@@ -0,0 +1,117 @@
+
+// ManifestHeader contains metadata for the manifest
+type ManifestHeader struct {
+	Format      uint16
```
Probably premature optimization to declare uint16 etc rather than uint.

icarus-sparry commented on src/swupd/manifest.go:
```diff
@@ -0,0 +1,117 @@
+	Version     uint32
+	Previous    uint32
+	FileCount   uint32
+	TimeStamp   uint64
```
It would be more normal to use time.Time for the timestamp. See [Fileinfo](https://golang.org/pkg/os/#FileInfo) from the standard library

icarus-sparry commented on src/swupd/manifest.go:
```diff
@@ -0,0 +1,117 @@
+	Previous    uint32
+	FileCount   uint32
+	TimeStamp   uint64
+	ContentSize uint64
```
The standard library uses int64 for the size.

icarus-sparry commented on src/swupd/manifest.go:
```diff
@@ -0,0 +1,117 @@
+	FileCount   uint32
+	TimeStamp   uint64
+	ContentSize uint64
+	Includes    []string
```
An array of strings? Can we not do better?

icarus-sparry commented on src/swupd/manifest.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// Read manifest file header
+func readManifestFileHeader(fields []string, m *Manifest) error {
```
Are you missing a loop here, to read all of the header, or is this badly named?
At the moment you ignore unknown headers, is that the intention?

matthewrsj replied:
This is poorly named. I haven't gotten to the code that will call this in a loop over the header lines.

matthewrsj replied:
Yes, the intention was to ignore unknown headers.

matthewrsj committed:
Add coverage.out to gitignore now we have tests

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj commented:
@cmarcelo, @icarus-sparry @phmccarty @tmarcu this is the PR I was talking about

matthewrsj committed:
Add remaining manifest read functions

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj committed:
Add tests for manifest read functions

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj commented:
This is ready to be reviewed again. I believe I've addressed @cmarcelo's comments

cmarcelo commented on src/swupd/manifest.go:
```diff
@@ -0,0 +1,118 @@
+			return err
+		}
+		// parsed is already uint64
+		m.Header.ContentSize = int64(parsed)
```
One idea: use ParseInt and verify that the value is not negative. That way you don't need to care about overflow here.

cmarcelo commented on src/swupd/manifest.go:
```diff
@@ -0,0 +1,118 @@
+	FileCount   int32
+	TimeStamp   time.Time
+	ContentSize int64
+	Includes    []Manifest
```
These should be pointers to the Manifest structures. Otherwise we'll have to copy definitions around, imagine parsing os-core-update, that includes os-core previously parsed.

I would actually just keep the names of the manifests being included and let a later step that needs resolve into the actual references (i.e. Includes would be []string or a type based on string).

matthewrsj replied:
Since at this point the only information we have about the includes is the name, I agree that this makes more sense as a []string. @icarus-sparry you were against this initially, did you think this should be a *[]Manifest?

icarus-sparry replied:
@cmarcelo is correct, we do want pointers to Manifests for the reasons he mentions.

We probably want a new type here, a structure with a name and a pointer to a manifest, the latter being filled in in the "resolve" step mentioned.

cmarcelo commented on src/swupd/manifest.go:
```diff
@@ -0,0 +1,118 @@
+	Files        []File
+	AddedCount   uint32
+	RemovedCount uint32
+	ChangedCount uint32
```
Minor: we talked about getting rid of those Counters, but it is fine if we do later on.

matthewrsj replied:
Oops, yes.

icarus-sparry replied:
If we have them instead as arrays of pointers to interned files (added files. removed files etc), then we always have len() which is a VERY cheap operation to get this data anyhow.

I think the point I made about this being the _internal_ (in memory) representation rather than the on disk version is worth repeating.

matthewrsj replied:
Except we still need to check more than just the len in the case where then length stays the same but we have several changed files.

cmarcelo replied:
My suggestion: remove the counters, don't add anything else. When the code that actually uses that information (added/etc) is added, it adds the necessary information (e.g. arrays of pointers, etc).

cmarcelo commented on src/swupd/manifest_test.go:
```diff
@@ -180,3 +180,81 @@ func TestReadManifestFileEntry(t *testing.T) {
+}
+
+func TestReadManifestFromFile(t *testing.T) {
+	t.Run("good manifest", func(t *testing.T) {
```
If the tests are completely separated like here, just create separated test functions (no need to t.Run in that case).

cmarcelo commented on src/swupd/manifest_test.go:
```diff
@@ -180,3 +180,81 @@ func TestReadManifestFileEntry(t *testing.T) {
+
+func TestReadManifestFromFile(t *testing.T) {
+	t.Run("good manifest", func(t *testing.T) {
+		path := "testfiles/manifest.good"
```
The convention is to use testdata/ directory instead of testfiles/. Go tool will not walk through that directory when looking for sources (not much practical difference for us, but still).

cmarcelo commented:
LGTM with comments.

I still think we can tweak things so manifest files with repeated "version" fields (or other unique field) could fail, but this can be done later on.

matthewrsj committed:
Make new swupd code gometalinter clean

This commit fixes several style issues identified by gometalinter.v1

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj committed:
Add Hashes and Pack structures

Include a very simple function to intern hashes so they can be stored on
the File as integers for faster comparison times. This function could be
made more efficient in the future with the use of a balanced tree, for
example.

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj committed:
Add tests for hash interning

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

Commit:
Different hash implementation

The point of using a map (hash table, associative array) to hold the
SHA values is to improve speed. Having an O(n^2) constructor is fine
for a test case of half a dozen unique values, but sucks if we have
half a million.

Also added a String method to allow values to be printed.

Signed-off-by: Icarus Sparry <icarus.w.sparry@intel.com>

bryteise commented on src/swupd/hash.go:
```diff
@@ -1,17 +1,23 @@
+}
+
+func (h hashval) String() string {
+	return *Hashes[int(h)]
```
I think a range check here to avoid runtime panics would be a good add.

icarus-sparry replied:
What can we do if a range check fails?

Since the point of interning the values is that we will never call delete on them, if a range check fails then it shows either a bug in the code or someone has cast a value to a hashval. Either way the only sensible thing to do is abort the program, and at least a runtime panic gives you some useful information.

bryteise replied:
I agree that we should fail, I just wasn't sure if a panic failure would be the best thing for this case. If you think panic is the right way to go that's fine but I usually prefer to avoid panics unless it is a 100% bug situation.

icarus-sparry replied:
I agree 100%. As I see it, we have to return a string. The only thing we could do differently is to say return fmt.Sprintf("Invalid hashval %d",int(h)), if the range check failed, but I think this will just make it too easy to ignore the problem. I think of the runtime panic being closer to an assert in C. if it fails we produce a corefile (which we then throw away so we can't debug the issue...)

bryteise replied:
Okay great, thanks for the explanation!

bryteise commented on src/swupd/hash_test.go:
```diff
@@ -27,3 +30,22 @@ func TestInternHash(t *testing.T) {
+	s[0] = '1'
+	sout := fmt.Sprintf("%v", v)
+	if sout == string(s) {
+		t.Errorf("in and out of hashtable do not match\n\t%v\n\t%v", sout, s)
```
Error message should say "do match" here I think.

icarus-sparry replied:
Yes.
Actually this test is pretty bogus.

bryteise commented on src/swupd/files_test.go:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	// space is never valid
+	t.Run("' '", func(t *testing.T) {
```
I'm confused how this isn't failing as the value being passed is a string and not a byte (should just be `' '` I think).

Commit:
Add trivial HashEquals function and tests

Wrote the test in a slightly unusual way to point out one of go's
gotcha features, that range produces a copy of the value rather than a
pointer to the value.

Wrapped one existing line to reduce maximum line length.

Signed-off-by: Icarus Sparry <icarus.w.sparry@intel.com>

matthewrsj commented:
If this is a decent base I'd like to merge it so we can actually work against master instead of creating PRs against a PR. @icarus-sparry @phmccarty @tmarcu @bryteise @cmarcelo

matthewrsj committed:
Add hash and flag string getters for Files

Add tests for these new functions as well.

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

tmarcu commented on src/swupd/hash.go:
```diff
@@ -0,0 +1,28 @@
+package swupd
+
+type hashval int
```
I really don't agree with aliasing/typedef-ing primitive types unless we are doing obscure, object specific things in functions with them. I'm all for defining new types where necessary but this can go overboard easily imo and I want to point it out now... i.e the comment doesn't make sense  when it says it returns an index (simple number) but the func returns a type hashval instead, just return an int.

tmarcu commented on src/swupd/files.go:
```diff
@@ -0,0 +1,203 @@
+	DeltaPeer *File
+}
+
+// typeFromFlag return file type based on flag byte
```
s/return/returns/g

tmarcu replied:
ah I see, return works too. English...

bryteise commented on src/swupd/files_test.go:
```diff
@@ -218,3 +219,34 @@ func TestSetHashInvalid(t *testing.T) {
+	}
+}
+
+func TestGetFlagString(t *testing.T) {
```
I'd expect another test with no flags set on the file to validate an error return.

matthewrsj replied:
Added

matthewrsj commented:
Merging per discussion in meeting. Additional PRs will be opened against master branch.
